File: p3.js
// ## Problem3
// Replace Words (https://leetcode.com/problems/replace-words/)

// Input: dict = ["cat","bat","rat"], sentence = "the cattle was rattled by the battery"
// Output: "the cat was rat by the bat"

// time: O(nk + lk)
// space: O(n)


/**
 * @param {string[]} dict
 * @param {string} sentence
 * @return {string}
 */
var alphabet = { 
    'a': 0, 
    'b': 1,
    'c': 2,
    'd': 3,
    'e': 4,
    'f': 5,
    'g': 6,
    'h': 7,
    'i': 8,
    'j': 9,
    'k': 10,
    'l': 11,
    'm': 12,
    'n': 13,
    'o': 14,
    'p': 15,
    'q': 16,
    'r': 17,
    's': 18,
    't': 19,
    'u': 20,
    'v': 21,
    'w': 22,
    'x': 23,
    'y': 24,
    'z': 25
}
class TrieNode {
    word;
    children = [];   
}
let root = new TrieNode();
function insert(word) {
    let cur = root;
    let char;
    for(let i = 0; i < word.length; i++) {
        char = word[i];
        if(!cur.children[alphabet[char]]) cur.children[alphabet[char]] = new TrieNode();
        cur = cur.children[alphabet[char]];
    }
    cur.word = word;
}

const replaceWords = function(dict, sentence) {
    for(let str of dict) { //nk + lk
        insert(str);
    }
    let sentenceArr = sentence.split(' '); 
    let originalArr = sentence.split(' '); 
    for(let k = 0; k < sentenceArr.length; k++) {
        let cur = root;
        for(let j = 0; j < sentenceArr[k].length; j++) {
            let char = sentenceArr[k][j];
            if(!cur.children[alphabet[char]] || cur.word) break;
            cur = cur.children[alphabet[char]];
        }
        if(cur.word) {
            sentenceArr[k] = cur.word;
        }
    }
    return sentenceArr.join(' ');
}

File: p3.java
// Time Complexity :O(ml + nl) where m is the number of words in dictionary and n is the number of words in sentence and l is the average length of those words
// Space Complexity :O(ml + nl)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach

class Solution {
    //Creating class for trienode consisting of isEnd and children
    class TrieNode{
        boolean isEnd;
        TrieNode[] children;
        public TrieNode(){
            isEnd = false;
            children = new TrieNode[26];
        }
    }
    TrieNode root;
   //To insert a word, iterate over the characters of the word while checking if it exists in the trie or not. If it does not exixt, iniatilze and insert it in trie. In the end mark last node's isEnd to true
    public void insert(String word){
        TrieNode curr = this.root;
        for(int i=0; i<word.length(); i++){
            char c = word.charAt(i);
            if(curr.children[c-'a'] == null){
                curr.children[c-'a'] = new TrieNode();
                
            }
            curr = curr.children[c-'a'];
        }
        curr.isEnd = true;
    }
   ///Iterate over the word and if you find null in the trie, return the word. if you find isEnd to be true in the trie, return the word till that node. In the end return the word
    public String search(String word){
        TrieNode curr = this.root;
        StringBuilder sbb = new StringBuilder();
        for(int i=0; i<word.length(); i++){
            char c = word.charAt(i);
            
            
            if(curr.isEnd){
                return sbb.toString();
            }
            sbb.append(c);
            if(curr.children[c-'a'] == null){
                
                return word;
                
            }
            
            curr = curr.children[c-'a'];
        }
        return word;
        
    }
    public String replaceWords(List<String> dictionary, String sentence) {
        this.root = new TrieNode();
        //Put words of dictionary in trie
        for(String word: dictionary){
            insert(word);
            
        }
        String[] strArray = sentence.split(" ");
        StringBuilder sb = new StringBuilder();
//Check for replacement words 
        for(int i=0; i<strArray.length; i++){
            
            String word = strArray[i];
            if(i!=0) sb.append(" ");
            sb.append(search(word));
        }

        return sb.toString();
    }
}